### DISCUSSION TITLE
RFC: NGql Filter pushed down to the storaged

### DISCUSSION THREAD

**Discussion by zhaojunnana** [vesoft-inc/nebula#4787] [General]
**Background**:

Some super nodes in business use need to be filtered through attributes. Because the filtering does not push down the large storage layer, the query times out. Forcibly truncating the storage layer through limit [] will result in data loss. In some aspects, whether and how much filtering can be pushed down directly affects the performance of the entire database.

In the nebula source code, the push-down logic is in the optimizer. When the filter encounters LeftJoin and InnerJoin, it can't push down, but normal multi-hop query statements will trigger these two behaviors.

The main reason why InnerJoin cannot be pushed down is that the storage layer does not support variable filtering, and the LeftJoin may have DestExpression. If these two problems are solved, you can basically push them down.

**General idea**: 
- build the GetNeighbors execution plan phase and prune the DestExpression filter statements; 
- With variable expressions, variables are transferred to the storage layer along with vid, and the storage layer FilterNode provides variable support.

**Reference PR**:https://github.com/vesoft-inc/nebula/pull/4796

**Comment by xtcyclist** [comment]
Thanks for posting this RFC. The general idea looks good to me. Link a detailed plan here: https://github.com/zhaojunnana/nebula/blob/master/docs/ngql_push_filter_down.md

**Comment by xtcyclist** [comment]
Regarding of further optimizing the processing of super nodes, I'm wondering whether filtering more properties direcly in the storaged, as proposed in this issue, would be sufficient for your applications in your business.

Although pushing down filters alleviate network transfers and the pressure on serialization and deserialization, the overhead of scanning the neighbourhood of super nodes is still there. A candidate for the worst case could be multiple concurrent queries scanning the same set of big super nodes.

Similar to what have been applied in RDBMS, we could try implementing shared node/edge scans in the storaged that serve the lookup requests for multiple concurrent queries to amortize the cost.

**Comment by zhaojunnana** [comment]
完全依靠属性过滤解决超级节点确实是不可行的，业务使用中建模方面是有节点打散处理的，超级节点可以理解为可以通过属性区分的部分大节点，下推过滤后我们的业务目前可以支撑起来。
正如您所说，节点过大就算下推计算到存储层，扫描点边的时间也不会有太多改善，只是减少了数据传输。计算下推之后，最大的问题就是扫描节点/边的耗时，不知道能不能把扫描工作分批进行，开多线程并发扫描，这种的前提是设计的rocksdb的key可以支持一定顺序的分批查询。这部分我还没有详细看过，只是初步有个想法。

**Reply by xtcyclist** [reply]
感谢介绍。明白了。

节点打散确实有用，最终还是需要从内核里面下功夫。在存储层或者存储引擎内部，确实可以通过一些线程级并发或者其它加速技术来降低开销，只要底层 I/O 资源还没有饱和；或者考虑在 storaged 层给超级节点添加 DRAM cache。不过从计算存储分离的整体架构出发，可能不宜给 storaged 预期太多的计算资源，比如 CPU 数、线程数和内存。可能也可以考虑通过在 graphd 和 storaged 之间进行流水线执行的方式，隐藏一部分底层的开销并提高网络资源的整体利用率；或者考虑通过共享 scan 的方式，在一段时间内提高 storaged 的资源利用率，提高它的整体吞吐，但不着重提高单个 query 的吞吐。

这部分问题需要考虑的因素还比较多，需要综合权衡一下，然后在未来的迭代中解决。我觉得回头我们可以在社区多多讨论这部分的场景和问题。

------------------

Thanks for the introduction on your practice of spliting a super node into a set of smaller nodes with shared properties.

It is indeed a good practice, especially when the kernel itself is not efficient enough here. Inside the storaged, indeed more threads could be utilized to accelerate the scanning of data, if the I/O is  not already saturated. Or, we could try caching super nodes in the DRAM of the storaged. However, to comply with the design goals of the overall architecture that seperates computation and storage, we probabaly shouldn't provision too much computation resources for the storaged, such as CPUs, threads, main memory spaces, etc. We probabaly need to consider other designs like pipelining the scanning of data in the storaged and the processing of them in the graphd to hide latencies and improve the utilization of network transfers. Or, we could try improving the resource utilizations of the storaged in a period of time to improve the aggregated query throughput via shared scanns.

There seems to be multiple issues that need to be discusssed and finally addressed in future developments. Looking forward to talk more on this in the community.

**Reply by zhaojunnana** [reply]
想到一起去了，流的方式也是我们理想中最好的方式。我们想象中最佳的实现：查询达到graphd后分成小批次，再大胆点就是graphd变成有状态集群，把小批次均分到graphd集群，充分利用多节点的算力。这也基本上能够杜绝graphd单节点面对高并发和大查询的oom情况。但这难度确实还是相当大，或许后面可以先考虑下单节点的流交互方式。

**Reply by xtcyclist** [reply]
👍 挺有道理的，在当前架构下持续深挖是比较靠谱的路线，挖掘一些还没有被开发出来的价值。后面可以多多交流。期待你的 PR。

**Reply by wenhaocs** [reply]
不错的讨论！其实“把graphd变成有状态集群，把小批次均分到graphd集群”也是我们在云上遇到问题的可行解决方案。在当前架构下，我们无法真正做到auto scale，没法利用serverless的能力。但从另一个角度说，我们最初的设计基于TP的，认为每个query应该很快就能完成。但随着使用，发现有时候很难区分是tp还是ap的query。如果要真正做到自适应tp/ap，以上讨论的方案都是值得拿出来仔细研究的。欢迎一起讨论！
